### PR TITLE
Add support for Bemis-Murcko molecular scaffold split for this WNV+flaviviral data (as used by molhiv dataset from MoleculeNet team, using DeepChem ScaffoldSplitter), reset hyperparameters to produce very low parameter count since active data is so sparse in AID 577, add support for AID 588689 which has ~10x the count of active molecules (similarly low fraction of active but higher absolute count)

### PR DESCRIPTION
**Pull request by willy-b**
**Adding support for Murcko scaffold splitting and switching the default dataset to AID 588689 from AID 577 since AID 577 has just 119 active molecules in total (after splitting this results in insufficient data, exacerbated in the scaffold split case vs random split).**

See https://deepchem.readthedocs.io/en/2.8.0/api_reference/splitters.html#scaffoldsplitter and 

"MoleculeNet: a benchmark for molecular machine learning" ( https://pubs.rsc.org/en/content/articlehtml/2018/sc/c7sc02664a ) section 3.2 Data Splitting "Scaffold splitting splits the samples based on their two-dimensional structural frameworks [...] MoleculeNet contributes the code for these splitting methods into DeepChem. " 

**and regarding use of scaffold split for their HIV antiviral dataset, section 3.1.9 

"As we are more interested in discover new categories of HIV inhibitors, scaffold splitting (introduced in the next subsection) is recommended for this dataset."**

(Scaffold split data is MORE DIFFICULT than randomly split data.)

Note we are NOT using MoleculeNet data, trying to prepare a similar dataset for WNV from publicly available data.

Also log average precision everywhere (e.g. during training) even when it is not main metric (here main metric is  ROCAUC).

(P.S. nobody yet has found an effective small molecule antiviral for WNV that has been approved, PCBA AID 577 dataset above is from 2006 and nobody found much from such datasets yet, set expectations accordingly. THIS IS NOT FOR A COMPETITION, unlike the ogbg-molhiv, OGB competitions do NOT feature WNV data, I am practicing to create a new dataset for applying the GNNs to WNV, maybe useful to contribute back to OGB later.)

### CODE REVIEW

**Comment by willy-b** [comment]
Adding support for https://pubchem.ncbi.nlm.nih.gov/bioassay/588689 which has over 1,000 active molecules instead of 119 in total like AID 577, we should be able to model the scaffold split of that data much better.

**Comment by willy-b** [comment]
Force pushed to consolidate multiple commits for enabling AID 588689 , running train/validation run on those with a very tiny 1,817 parameter model (not necessarily the optimal but a parsimonious starting point), will post shortly, thankfully looks much much better than what I was seeing with the AID 577 low data (~1 dozen active molecules in validation split, < 100 molecules in train split) case.

**Comment by willy-b** [comment]
Preview of exciting **preliminary results (seed 1 for weight initialization/training data sorting, validation split, seed 0 / default for data permutation before scaffold split)** for the Flavivirus Genome Capping Enzyme Inhibition dataset AID 588689 with a **structural scaffold split per the MoleculeNet approach using DeepChem** (so validation set has different scaffolds than training):

Flavivirus_Genome_Capping_Enzyme_Inhibition_Prediction_using_2-hop_GIN_hidden_dim_8_and_GraphNorm_ROC_CURVE_valid
Flavivirus_Genome_Capping_Enzyme_Inhibition_Prediction_using_2-hop_GIN_hidden_dim_8_and_GraphNorm_PRC_CURVE_valid

This is similar to what I am seeing on the other seeds (I am running n=10 with gradient clipping and without gradient clipping for comparison).

Note that the AID 588689 is not strictly a WNV dataset (West Nile Virus is a flavivirus) but it is supposed to be predictive of generic Flavivirus Genome Capping Enzyme Inhibition (but they tested using the Dengue capping enzyme).

Note that on this seed the average precision was 11% , when the expected by chance was < 0.4%, but also that this turns out to be one of the best and more typical is 7-10% AP (still much higher than chance <0.4%). Colab notebooks will be linked soon.

**Comment by willy-b** [comment]
**VERY PRELIMINARY, VALIDATION SET PERFORMANCE ONLY, EVENTUAL TEST SET PERFORMANCE MAY BE WORSE** (though using a low number of parameters makes it less likely and validation set here is a scaffold split so it is already generalizing)

tiny-GIN-for-WNV - train/validation check of AID 588689 with 2 layer 8d network (1,817 parameters), 20 epochs; **SCAFFOLD SPLIT**, same scaffold split (seed 0 / default; note there are many scaffolds with relatively small numbers of molecules so the assignment of scaffolds to training/validation/test is still random), N=10 seeds for random weight initialization / training data order on each run:

https://colab.research.google.com/drive/1cwPXmJLkonOpW3aw6Iyjk_2vEXPfKLPG

```
Validation set ROCAUC 89.7% +/- 1.0% (mean +/- std; n=10)
95% confidence interval for sample mean of n=10 is 89.1 to 90.3%

Validation set Average Precision 8.4% +/- 1.7% (mean +/- std; n=10),
95% confidence interval for sample mean of n=10 is 7.3 to 9.4%
```

Not claimed to be optimal hyperparameters, just a parsimonious starting point. I will be running train/validation with other hyperparameters on the scaffold split to compare.

Note early stopping was used in the sense that out of the N epochs, the best model by validation was kept, which can distort the validation score higher (one of the reasons why eventually a separate holdout test split will be reported), though in this case it was close to the last epochs value in all cases so very little distortion in this particular case (pretty smooth improvement across epochs, not randomly jumping around).

Gradient norm clipping is not essential (had been added earlier in this same PR only to stabilize training on the tiny dataset AID 577 when using very few parameters based on the idea of balancing the parameter changes with the weight decay more closely)  for AID 588689 where we have more data, but it did produce a trend towards higher average precision. In general adding another hyperparameter to tune (though I used same value I found on AID 577 without adjustment here) is something to be avoided unless it is a big deal so will consider to remove gradient norm clipping as an option (same as before) before merging this PR, as results are roughly the same without (very slightly lower; I ran a separate notebook with gradient norm clipping effectively turned off by setting the limit orders of magnitude higher).

**Comment by willy-b** [comment]
The same 32K parameter model configuration / hyperparameters used in my 2024-07 leaderboard submission to OGB for the ogbg-molhiv works fine for AID 588689 out of the box (could not do a hyperparameter search and just reuse it there - I started tuning because AID 577 which only had 119 active molecules in total, a tiny dataset, required a different approach.)

Same model configuration / hyperparameters used in my 2024-07 leaderboard submission to OGB for the ogbg-molhiv but for AID 588689 "Primary and Confirmatory Screening for Flavivirus Genomic Capping Enzyme Inhibition", and with GraphNorm enabled for a slight boost:

 https://colab.research.google.com/drive/17JILSAkHj4IvKIpksBpjD0yrSyIVGZTv

```
32,449 parameter model, using 9 atom features only
2-layer 64-hidden dimension GIN with post-graph embed MLP, batch size 128, 1e-6 weight decay, and GraphNorm:

Validation set ROCAUC 94.1% +/- 0.2% (mean +/- std; n=10)
Validation set ROCAUC 95% confidence interval for sample mean at n=10 is 94.0 to 94.2%
Validation set Average Precision (AP) 19.0% +/- 1.7% (mean +/- std; n=10) (chance AP level would be 0.38%)
Validation set Average Precision (AP) 95% confidence interval for sample mean at n=10 is 17.9 to 20.0%
```

**Comment by willy-b** [comment]
Representative Receiver Operating Characteristic (ROC) and Precision-Recall Curve (PRC) figures for the previous comment ( https://github.com/willy-b/tiny-GIN-for-WNV/pull/14#issuecomment-2566594352 ), seed 0 from https://colab.research.google.com/drive/17JILSAkHj4IvKIpksBpjD0yrSyIVGZTv

32,449 parameter model, using 9 atom features only
2-layer 64-hidden dimension GIN with post-graph embed MLP, batch size 128, 1e-6 weight decay, and GraphNorm

Flavivirus_Genome_Capping_Enzyme_Inhibition_Prediction_using_2-hop_GIN_hidden_dim_64_and_GraphNorm_ROC_CURVE_valid

Flavivirus_Genome_Capping_Enzyme_Inhibition_Prediction_using_2-hop_GIN_hidden_dim_64_and_GraphNorm_PRC_CURVE_valid

Separately, coming up next I am currently running the last day the OGB team baselines which will take some time to generate 10 seeds as they are millions of parameters models with a higher number of training epochs, currently running one at a time on an A100 (my tiny GINs above were run on Google Colab CPU instance and 10 training runs can be done very quickly especially for the sub 2K parameter model), but I am excited to compare.
